### DISCUSSION TITLE
What would be the correct approach

### DISCUSSION THREAD

**Discussion by Lockszmith-GH** [AdguardTeam/AdGuardHome#3581] [Q&A]
I've been trying to *properly* setup my internal DNS using AdGuard Home, and I got everything working, but it looks to me as if the solution should have been much simpler/cleaner, instead I got it working with a bunch of 'workarounds'.

I'll lay out my topology and current setup and then I'll ask my questions, next to relevant configuration lines I added a `[#]` letter for later reference.

## Topology
I have a setup that is as follows:
```
External DNS:
example.com A record IPv4 to my LAN
*.example.com CNAME to example.com
c.example.com A record IPv4 someplace else
d.example.com A record IPv4 someplace else

LAN IPs
hostname: a.lan | IPv$: 192.168.100.1
hostname: b.lan | IPv$: 192.168.100.1
```

My goal is to have `(*.)example.com` be answered with `a.lan`'s address, except for `b.example.com` which will be answered by `b.lan` and `c.example.com` and `d.example.com` that will be answered by the external addresses.

## Current Setup
> ### In *AdGuard Home* -> *Filters* -> ***DNS rewrite***:
>> ```yaml
>> rewrites:
>>   - domain: a.lan
>>     answer: 192.168.100.1
>>   - domain: b.lan
>>     answer: 192.168.100.2
>>   - domain: b.lksz.me
>>     answer: b.lan [A]
>> ```
>
> ### In *AdGuard Home* -> *Filters* -> ***Custom filtering rules***:
>> ```adblock
>> *.example.com^$dnsrewrite=NOERROR;A;192.168.100.1,denyallow=c.example.com|d.example.com [B]
>> ```

### My Questions
I would really love to have everything defined in a single place, preferably the *easily editable* ***Custom filtering rules***, but it seems that the internal CNAME rewrite (see `[A]`) doesn't work, and I needed to use an A record rewrite instead (see `[B]`).

1. Is this possible at all (to have all the configuration in one place)?

2. I think the one thing that really prevents me from achieveing this goal, is the fact that a `$dnsrewrite` cannot use a `HOSTS` type entry or another `$dnsrewrite` as a CNAME destination.
   What I mean is, that the following **does NOT work**:
   ```
   *.example.com^$dnsrewrite=NOERROR;CNAME;a.lan,denyallow=c.example.com|d.example.com [B]
   ```

    Anyone know if this is this a bug? design flaw? or a design choice?

**Comment by ameshkov** [comment]
> is the fact that a $dnsrewrite cannot use a HOSTS type entry or another $dnsrewrite as a CNAME destination

We had some discussions before on this topic and decided not to overcomplicate $dnsrewrite since making it recursive is a bit risky.

@ainar-g could you please remind the past discussions? It seems there are more than one person bothered with the current design choice, we may want to reconsider.

**Reply by ainar-g** [reply]
@ameshkov, partially in #3350 and it also largely affects #2499.

~~I told you that this will come up again :stuck_out_tongue_winking_eye:.~~

**Reply by ameshkov** [reply]
Well, #3350 is a little different story, AGH is not a recursive resolver and it shouldn't follow the CNAMEs chain.

@ainar-g shall we merge all this into a single issue? Probably, this should all be merged into #2499 since the current $dnsrewrite implementation may not be fully compatible with the old DNS rewrites.

**Reply by ainar-g** [reply]
@ameshkov, I mean, #3350 is basically a blocker of #2499.  But yeah, I do think that we could merge them.

And while we're at it, I think we should postpone the merger of all these rewrites into one until the v0.108 cycle.  (The merger of `/etc/hosts`-style rewrites and `$dnsrewrite` can probably still land in v0.107.)

**Reply by Lockszmith-GH** [reply]
How about a simplified approach, instead of supporting full blown recursion, leave CNAME support to be an external one, but add a keyword for internally defined names (let's say HOSTS)?

This way you not only solve this particular issue, you also document the reasoning of non-recursion in documentation, yet still provide a solution for a 'simple' recursive-like situation.

**Reply by ameshkov** [reply]
I am okay with `$dnsrewrite` following rewrites "internally" (i.e. following other $dnsrewrite rules) since this is anyway necessary for implementing #2499. As I understand, this covers your use case in particular?

**Reply by Lockszmith-GH** [reply]
oh yes, it does.

And I understand the fear of loops being defined.
Will the order in which the rules are presented have any weight. (more of a thinking out loud question, nothing relevant right now)

**Reply by Lockszmith-GH** [reply]
13 months later and not much activity on #2499 nor #3350. Are there any updates?
